### ISSUE TITLE
How to use different embedding layer than training in evaluation

### ISSUE THREAD

**Issue by jyori112** [allenai/allennlp#2640]
Please first search our GitHub repository for similar questions.  If you don't find a similar example you can use the following template:

**System (please complete the following information):**
 - OS: Ubuntu 16
 - Python version: 3.7.1
 - AllenNLP version: 0.8.2
 - PyTorch version: 1.0.1.post2

**Question**

Hello, 

I am trying to develop a simple multilingual model by using bilingual word embeddings as the embedding layer.
So, I train the model in English fixing the embedding layer to be the bilingual word embeddings, then I want to replace it with German word embeddings in the evaluation.

I am struggling to find the best way to do this.

I tried the override option (`-o`) to set `pretrained_file` for token-embedder module, but it did not work.

Going through the code, I found `remove_pretrained_embedding_params` function at https://github.com/allenai/allennlp/blob/12626ac4ad8b290d1826d2f194b7e80e582a2d82/allennlp/models/model.py#L358 which is called at https://github.com/allenai/allennlp/blob/12626ac4ad8b290d1826d2f194b7e80e582a2d82/allennlp/models/model.py#L266.
These codes, I believe, remove any configuration of pretrained_file in the evaluation.

I hope that I can get some help here.

Thank you.

**Comment by matt-gardner** [comment]
So, in allennlp you really just want this to be a single "embedding layer" that has both languages in its vocabulary.  That's the easiest way to handle this.  One option for doing this is to compute the vocabulary using the test data (there are configuration options to the trainer that let you do this) - this will get the German words loaded, with their embeddings, which won't actually be used until you see those words at test time (unless you have a softmax over the vocabulary at some point in your model, then this could mess with gradients a bit).  Another option is to expand your vocabulary at test time (the `evaluate` command has an option for this) - this will do essentially the same thing as the first option, but not require you to know the test dataset at training time.

Both of these options will do funny things if there are overlapping words in the two languages - you will only get one vector for each unique token.  If this is an issue, you could prepend each token with a language identifier.

**Comment by jyori112** [comment]
Thank you for the feedback.

I believe a single embedding layer would be OK for most of the case.
However, given how my project is organized, I feel it would mess up my code.  So, I decided to create my own `Embedding` class with an additional feature. 

I still believe the overriding embedding layer is a nice feature to have.

Thank you

**Comment by vidurj** [comment]
If you think the problem is not solved, feel free to open another issue.

**Comment by mayhewsw** [comment]
Correct me if I'm wrong, but in order to implement @matt-gardner's suggestion, you have to combine two vector files into one, which you then give as the `pretrained_file` path in the config. Here's a gist that does this: https://gist.github.com/mayhewsw/3f564aa268c62189c2f1a37c0654ad90

**Comment by matt-gardner** [comment]
@mayhewsw, hmm, yes, I believe that's true, thanks for catching that.

**Comment by mayhewsw** [comment]
For the sake of posterity: I've tried both methods (combine vectors, using my gist; and extend using `extend-vocab` and `embedding_sources` at evaluate time) and found that the extend method gives much higher scores than the combine method (doing cross-lingual NER). It's baffling, because they should both work. I haven't figured it out yet. It may have to do with vocabulary conflicts, where you have to choose one embedding over another, although this is primarily punctuation and numbers.

**Comment by matt-gardner** [comment]
If you have any kind of regularization during training, this could be impacting your unused embeddings.  That's my best guess as to the difference.
